// storybook-readme/env/react/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _StoryPreview = require('./components/StoryPreview');

var _StoryPreview2 = _interopRequireDefault(_StoryPreview);

var _FooterDocs = require('./components/FooterDocs');

var _FooterDocs2 = _interopRequireDefault(_FooterDocs);

var _common = require('../common');

var _common2 = _interopRequireDefault(_common);

var _highlite = require('../../services/highlite');

var _highlite2 = _interopRequireDefault(_highlite);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Story = function (_React$Component) {
  _inherits(Story, _React$Component);

  function Story() {
    _classCallCheck(this, Story);

    return _possibleConstructorReturn(this, (Story.__proto__ || Object.getPrototypeOf(Story)).apply(this, arguments));
  }

  _createClass(Story, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var el = _reactDom2.default.findDOMNode(this);

      (0, _highlite2.default)(el, {
        withJSX: true
      });
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      var el = _reactDom2.default.findDOMNode(this);

      (0, _highlite2.default)(el, {
        withJSX: true
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          storyFn = _props.storyFn,
          story = _props.story,
          kind = _props.kind,
          config = _props.config,
          _props$docs = _props.docs,
          docsAfterPreview = _props$docs.docsAfterPreview,
          docsBeforePreview = _props$docs.docsBeforePreview;


      var PreviewComponent = config.PreviewComponent || _StoryPreview2.default;
      var FooterComponent = config.FooterComponent || _FooterDocs2.default;

      return _react2.default.createElement(
        'div',
        { style: { padding: '10px' } },
        docsBeforePreview && docsBeforePreview.map(function (doc, index) {
          return _react2.default.createElement('div', {
            key: index,
            className: 'markdown-body',
            dangerouslySetInnerHTML: { __html: doc }
          });
        }),
        _react2.default.createElement(
          PreviewComponent,
          null,
          storyFn({ kind: kind, story: story })
        ),
        docsAfterPreview && docsAfterPreview.map(function (doc, index) {
          return _react2.default.createElement('div', {
            key: index,
            className: 'markdown-body',
            dangerouslySetInnerHTML: { __html: doc }
          });
        }),
        config.docsAtFooter && _react2.default.createElement(
          FooterComponent,
          null,
          _react2.default.createElement('div', {
            className: 'markdown-body',
            dangerouslySetInnerHTML: { __html: config.docsAtFooter }
          })
        )
      );
    }
  }]);

  return Story;
}(_react2.default.Component);

function withDocsCallAsHoc(_ref) {
  var docs = _ref.docs,
      config = _ref.config,
      storyFn = _ref.storyFn;

  return function (_ref2) {
    var kind = _ref2.kind,
        story = _ref2.story;
    return _react2.default.createElement(Story, {
      docs: docs,
      config: config,
      storyFn: storyFn,
      kind: kind,
      story: story
    });
  };
}

function withDocsCallAsDecorator(_ref3) {
  var docs = _ref3.docs,
      config = _ref3.config;

  return function (storyFn, _ref4) {
    var kind = _ref4.kind,
        story = _ref4.story;
    return _react2.default.createElement(Story, {
      docs: docs,
      config: config,
      storyFn: storyFn,
      kind: kind,
      story: story
    });
  };
}

function doc(_ref5) {
  var docs = _ref5.docs,
      config = _ref5.config;

  return function (_ref6) {
    var kind = _ref6.kind,
        story = _ref6.story;
    return _react2.default.createElement(Story, {
      docs: {},
      config: Object.assign({}, config, {
        PreviewComponent: function PreviewComponent(_ref7) {
          var children = _ref7.children;
          return _react2.default.createElement(
            'div',
            null,
            children
          );
        }
      }),
      storyFn: function storyFn() {
        return docs.map(function (doc, index) {
          return _react2.default.createElement('div', {
            key: index,
            className: 'markdown-body',
            dangerouslySetInnerHTML: { __html: doc }
          });
        });
      },
      kind: kind,
      story: story
    });
  };
}

exports.default = {
  doc: doc,
  withReadme: _common2.default.withReadme,
  withDocs: {
    callAsDecorator: withDocsCallAsDecorator,
    callAsHoc: withDocsCallAsHoc
  }
};
